docs: add README for blurhash_service (#3725)
divinevideo/divine-mobile: mobile/packages/blurhash_service/README.md

# blurhash_service

Status: Current
Validated against: `pubspec.yaml` on 2026-04-30.

> Purpose: blurhash generation and decoding for video thumbnail placeholders. Provides `BlurhashService`, `BlurhashData`, and `BlurhashCache`.

Reviewer: The exported public surface in `lib/blurhash_service.dart` is wider than this list — `BlurhashException` and `VineContentType` are also exported, and the latter is the parameter type for the public `getBlurhashForContentType` (lib/src/blurhash_service.dart:178). Worth listing so readers don't miss them.

```suggestion
Purpose: blurhash generation and decoding for video thumbnail placeholders. Provides `BlurhashService`, `BlurhashData`, `BlurhashCache`, `BlurhashException`, and the `VineContentType` enum consumed by `getBlurhashForContentType`.
```

Used by: the Flutter app to display progressive loading placeholders while video thumbnails are fetched.

## Behaviour

> - Encoding runs in a background isolate via `compute` to keep the UI thread free.

Reviewer: Strictly true for `generateBlurhash`, but `generateBlurhashFromImage` (lib/src/blurhash_service.dart:95-115) does `image.toByteData(format: ImageByteFormat.png)` on the caller's thread before handing off — PNG encoding of a `ui.Image` is non-trivial. Worth narrowing the claim or footnoting the `fromImage` variant so a future caller doesn't trust the blanket statement from a hot UI path.

```suggestion
- Encoding runs in a background isolate via `compute` to keep the UI thread free. (`generateBlurhashFromImage` performs the PNG conversion on the caller's thread before delegating — keep it off the UI path for large images.)
```

> - Component counts are chosen by aspect ratio: `4×7` for 9:16 portrait, `4×4` for 1:1 square.

Reviewer: The selector at `lib/src/blurhash_service.dart:64` returns `4×4` for any `ratio >= 0.9` — that's square **and all landscape**, not just 1:1. A 16:9 thumbnail gets `4×4` here, which the current wording would surprise a reader on.

```suggestion
- Component counts are chosen by aspect ratio: `4×7` for portrait (ratio < 0.9), `4×4` for square or landscape (ratio ≥ 0.9).
```

- Decoding returns a `BlurhashData` object with pixel data, representative colours, and a gradient fallback.

## Legacy hash compatibility

> Blurhashes stored in Nostr events published before PR #3684 were encoded with a 4:3 component ratio. They still render acceptably — colours are representative — but fine detail will not align with 9:16 content. This is an expected tradeoff for pre-existing events; no backfill was performed. New uploads generate hashes at the correct 9:16 dimensions.

Reviewer: Minor wording: #3684's actual change was `defaultComponentX = 4`, `defaultComponentY = 3` → aspect-ratio-aware. So the old hashes used a `4×3` **component grid** applied regardless of source aspect ratio, not "encoded for 4:3 source". Both readings land close, but being literal helps the next debugger.

```suggestion
Blurhashes stored in Nostr events published before PR #3684 were encoded with `4×3` components (the previous defaults), regardless of the source aspect ratio. They still render acceptably — colours are representative — but fine detail will not align with 9:16 content. This is an expected tradeoff for pre-existing events; no backfill was performed. New uploads generate hashes at the correct 9:16 dimensions.
```

# blurhash_service

Status: Current
Validated against: `pubspec.yaml` on 2026-04-30.

> Purpose: blurhash generation and decoding for video thumbnail placeholders. Provides `BlurhashService`, `BlurhashData`, `BlurhashCache`, `BlurhashException`, and the `VineContentType` enum consumed by `getBlurhashForContentType`.

Reviewer 2: Expanded the public-surface summary here so the README matches the actual package exports, including `BlurhashException` and `VineContentType`.

Used by: the Flutter app to display progressive loading placeholders while video thumbnails are fetched.

## Behaviour

> - Encoding runs in a background isolate via `compute` to keep the UI thread free. (`generateBlurhashFromImage` performs the PNG conversion on the caller's thread before delegating — keep it off the UI path for large images.)

Reviewer 2: This caveat matters: `generateBlurhash` is isolated, but `generateBlurhashFromImage` still does the PNG conversion on the caller thread first, so the README should not over-promise “UI-thread free” for that path.

> - Component counts are chosen by aspect ratio: `4×7` for portrait (ratio < 0.9), `4×4` for square or landscape (ratio ≥ 0.9).

Reviewer 2: Updated the aspect-ratio note to match the actual selector: anything with `ratio >= 0.9` gets the `4×4` grid, which includes square and landscape inputs.

- Decoding returns a `BlurhashData` object with pixel data, representative colours, and a gradient fallback.

> - `BlurhashCache` provides a bounded in-memory cache of decoded blurhash data with expiry-based eviction.

Reviewer 2: Added a brief cache note here so `BlurhashCache` is not just named in the exported surface without telling readers what it actually does.

## Legacy hash compatibility

> Blurhashes stored in Nostr events published before PR #3684 were encoded with `4×3` components (the previous defaults), regardless of the source aspect ratio. They still render acceptably — colours are representative — but fine detail will not align with 9:16 content. This is an expected tradeoff for pre-existing events; no backfill was performed. New uploads generate hashes at the correct 9:16 dimensions.

Reviewer 2: Clarified the legacy note so it refers to the old fixed `4×3` component grid rather than implying the source media itself was 4:3.
